wordpress-mobile/WordPress-iOS
# Widgets: Track loaded widgets when application is opened

Fixes #21699

Current Widget analytics are hard to interpret. We use separate events for each widget kind and add a number of different widget families to properties. We only track the event for the second time when properties change. The tracking of these events is also triggered unpredictably and driven by the system. 

I noticed from the experience that the way we track widgets makes it hard to understand both the overall usage of widgets as well the most popular widget types. This makes those events hardly useful.

## Changes

My goal for widget analytics is to understand two questions:
- The percentage of users who use any widgets
- The proportional usage of different widget kinds and families

## Possible solutions

### Track added and removed widgets

[As described in this article](https://tinyurl.com/ycktdz3a), we can use `getTimeline` and `getCurrentConfigurations` to keep track of previous and current widget configurations and try to track what type of widgets have been added and removed. 

I experimented with this approach. The drawback remains the unpredictable triggering of events. `getTimeline` is usually not triggered when the widgets are removed, so we could only reliably track added widgets. Also, we would still need to do additional calculations to determine the overall usage of widgets.

### Track all widgets when the application is opened

WooCommerce iOS tracks widgets when an application is opened (https://github.com/woocommerce/woocommerce-ios/pull/7717).

Although it does not allow to understand a more granular interaction with widgets, it does give a straightforward view into the proportional usage of different kinds of widgets (`?eventname=woocommerceios_application_opened ` ). Moreover, we can derive the overall usage of widgets by looking at how many application open events do not contain any widgets at all. 

This approach satisfies our current needs.

## Solution:

1. Created `WidgetAnalytics` with `trackLoadedWidgetsOnApplicationOpened` method which loads widget configuration and adds any different widgetConfig to "widgets" property of `widgets_loaded_on_application_opened` event
2. Calling the method when application is opened
3. Removed an older widget tracking code

I think this solution is much more simple to understand and interpret.

## To test:

1. Launch Jetpack app
2. Add home and/or lock screen widgets
3. Open app
5. Confirm that added widget types are tracked
6. Repeat steps 2-3 couple of times

I noticed that it can take `getCurrentConfigurations` a few seconds to return a newest configuration.

## Regression Notes
1. Potential unintended areas of impact

I removed existing widget tracks. I don't think anyone found them valuable but I could be wrong.

2. What I did to test those areas of impact (or what existing automated tests I relied on)


3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:


UI Changes testing checklist:
